File: frontend/types/capacitacion.ts
// 🔴 `entidad_capacitadora`, `modalidad` y `tipo` son string LIBRE, sin union de literales, a
// propósito: las columnas son text sin CHECK (mig 116) y el vocabulario real no se conoce — el
// Excel trae UN solo valor de cada una. Angostar el tipo acá inventaría un contrato que el
// backend y la base no tienen. Espejo del mismo comentario en schemas/capacitacion.py.
export interface Capacitacion {
  id: string
  empresa_id: string
  empresa_nombre: string | null
  nombre: string
  descripcion: string | null
  categoria: string | null
  duracion_horas: number | null
  entidad_capacitadora: string | null
  modalidad: string | null
  tipo: string | null
  obligatoria: boolean
  activo: boolean
  created_at: string
}

export interface CapacitacionCreate {
  empresa_id: string
  nombre: string
  descripcion?: string
  categoria?: string
  duracion_horas?: number
  entidad_capacitadora?: string
  modalidad?: string
  tipo?: string
  obligatoria: boolean
}

export interface CapacitacionUpdate {
  nombre?: string
  descripcion?: string
  categoria?: string
  duracion_horas?: number
  entidad_capacitadora?: string
  modalidad?: string
  tipo?: string
  obligatoria?: boolean
  activo?: boolean
}

export interface CapacitacionListResponse {
  items: Capacitacion[]
  total: number
}

export interface Asignacion {
  id: string
  empresa_id: string
  empresa_nombre: string | null
  capacitacion_id: string
  capacitacion_nombre: string | null
  // 🔴 null = fila de nombre libre (migración 116): la formación de alguien que el import no
  // matcheó contra el padrón, o de alguien que ya no trabaja acá. `nombre_libre` es el nombre
  // crudo del Excel y es lo único que identifica a esa persona.
  // `empleado_id === null` ES el "no está vinculado": no hay flag booleano aparte a propósito,
  // porque dos formas de decir lo mismo divergen (evaluaciones manda `asignado` Y `empleado_id`).
  empleado_id: string | null
  empleado_nombre: string | null
  nombre_libre: string | null
  area_id: string | null
  area_nombre: string | null
  // ⚠️ `anio` y `mes` son string, no number: las columnas son TEXT (mig 116) porque el Excel
  // los trae a mano y sin normalizar ("2026", "marzo"/"Marzo"). No parsearlos.
  proyecto: string | null
  anio: string | null
  mes: string | null
  estado: "pendiente" | "en_curso" | "completado"
  fecha_asignacion: string | null   // "YYYY-MM-DD"
  fecha_limite: string | null
  fecha_completado: string | null
  certificado_url: string | null    // storage path (privado) — usar signed URL para descargar
  created_at: string
}

export interface AsignacionCreate {
  capacitacion_id: string
  // ⚠️ Acá SIGUE siendo obligatorio, al revés que en `Asignacion`. Asignar un curso desde la
  // pantalla exige elegir un colaborador: el backend resuelve la empresa a partir de él
  // (`AsignacionCreate.empleado_id: UUID` en schemas/capacitacion.py) y sin él responde 422.
  // Las filas de nombre libre nacen del import, que no pasa por este endpoint.
  empleado_id: string
  fecha_asignacion?: string
  fecha_limite?: string
  proyecto?: string
  anio?: string
  mes?: string
  nombre_libre?: string
}

export interface AsignacionUpdate {
  estado?: "pendiente" | "en_curso" | "completado"
  fecha_limite?: string
  fecha_completado?: string
  proyecto?: string
  anio?: string
  mes?: string
  nombre_libre?: string
}

export interface AsignacionListResponse {
  items: Asignacion[]
  /**
   * 🔴 HOY `total === items.length` porque este listado NO pagina: el backend devuelve todo.
   * El día que pagine (sesiones 2–5), `total` pasa a ser "cuántas hay" y `items` una página.
   * Un contador escrito como `items.length` va a seguir compilando y va a decir 20 sobre 400.
   * Usá `total` para contar y `items` sólo para recorrer.
   */
  total: number
  page: number
  page_size: number
  total_pages: number
}

File: backend/schemas/capacitacion.py
"""
Schemas Pydantic para el módulo de capacitaciones.
Capacitacion: catálogo por empresa — empresa_id explícito en Create.
Asignacion: asignación empleado × curso — empresa_id heredado del empleado al crear.
"""
from datetime import date, datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel


# ── Catálogo de capacitaciones ─────────────────────────────────────────────────

# 🔴 `entidad_capacitadora`, `modalidad` y `tipo` van como `str` LIBRE, sin Literal ni Enum, a
# propósito: las columnas son text sin CHECK (mig 116) porque el Excel real es una planilla a
# mano y el vocabulario no se conoce todavía (42 filas traen UN solo valor de modalidad y de
# tipo). Angostar el conjunto acá inventaría un contrato que la base no tiene, contra un padrón
# que no alcanza para decidirlo. Cuando el vocabulario se estabilice, normalizar es una migración
# chica + un Literal; al revés no.

class CapacitacionCreate(BaseModel):
    empresa_id: UUID
    nombre: str
    descripcion: Optional[str] = None
    categoria: Optional[str] = None
    duracion_horas: Optional[float] = None
    entidad_capacitadora: Optional[str] = None
    modalidad: Optional[str] = None
    tipo: Optional[str] = None
    obligatoria: bool = False


class CapacitacionUpdate(BaseModel):
    nombre: Optional[str] = None
    descripcion: Optional[str] = None
    categoria: Optional[str] = None
    duracion_horas: Optional[float] = None
    entidad_capacitadora: Optional[str] = None
    modalidad: Optional[str] = None
    tipo: Optional[str] = None
    obligatoria: Optional[bool] = None
    activo: Optional[bool] = None


class CapacitacionResponse(BaseModel):
    id: str
    empresa_id: str
    empresa_nombre: Optional[str] = None
    nombre: str
    descripcion: Optional[str] = None
    categoria: Optional[str] = None
    duracion_horas: Optional[float] = None
    entidad_capacitadora: Optional[str] = None
    modalidad: Optional[str] = None
    tipo: Optional[str] = None
    obligatoria: bool
    activo: bool
    created_at: datetime


class CapacitacionListResponse(BaseModel):
    items: List[CapacitacionResponse]
    total: int


# ── Asignaciones empleado × capacitación ─────────────────────────────────────

# ⚠️ `anio` y `mes` son `str`, no int, y NO es un descuido: las columnas son TEXT (mig 116)
# porque el Excel los trae a mano y sin normalizar ("2026", "marzo"/"Marzo"). Tiparlos int acá
# inventaría un contrato que la base no tiene y rompería el import con el archivo real.
# `nombre_libre` en el Create no reemplaza a `empleado_id` (que sigue obligatorio: la pantalla
# asigna a un colaborador del padrón y la empresa se hereda de él): es el nombre crudo del Excel,
# y puede acompañar a una fila vinculada. Las filas SIN empleado nacen del import (A5.2), que
# escribe por el repo, no por este endpoint.

class AsignacionCreate(BaseModel):
    capacitacion_id: UUID
    empleado_id: UUID
    fecha_asignacion: Optional[date] = None
    fecha_limite: Optional[date] = None
    proyecto: Optional[str] = None
    anio: Optional[str] = None
    mes: Optional[str] = None
    nombre_libre: Optional[str] = None


class AsignacionUpdate(BaseModel):
    estado: Optional[str] = None
    fecha_limite: Optional[date] = None
    fecha_completado: Optional[date] = None
    proyecto: Optional[str] = None
    anio: Optional[str] = None
    mes: Optional[str] = None
    nombre_libre: Optional[str] = None


class AsignacionResponse(BaseModel):
    id: str
    empresa_id: str
    empresa_nombre: Optional[str] = None
    capacitacion_id: str
    capacitacion_nombre: Optional[str] = None
    # 🔴 `empleado_id` es OPCIONAL desde la migración 116: una fila de nombre libre no cuelga de
    # ningún colaborador del sistema. Con `str` a secas, Pydantic tira ValidationError al armar
    # la respuesta y el endpoint devuelve 500. `nombre_libre` es el nombre crudo del Excel de
    # formación, y es lo ÚNICO que identifica a esa persona: sin él la fila sale anónima.
    empleado_id: Optional[str] = None
    empleado_nombre: Optional[str] = None
    nombre_libre: Optional[str] = None
    area_id: Optional[str] = None
    area_nombre: Optional[str] = None
    proyecto: Optional[str] = None
    anio: Optional[str] = None
    mes: Optional[str] = None
    estado: str
    fecha_asignacion: Optional[date] = None
    fecha_limite: Optional[date] = None
    fecha_completado: Optional[date] = None
    certificado_url: Optional[str] = None
    created_at: datetime


class AsignacionListResponse(BaseModel):
    items: List[AsignacionResponse]
    # 🔴 HOY `total == len(items)` PORQUE ESTE LISTADO NO PAGINA: `asignacion_repo.find_all` trae
    # todas las asignaciones del filtro. Es el más grande de los cinco (1.558 filas en la base de
    # escala), o sea el primero al que se le va a notar.
    # 🔴 EL DÍA QUE PAGINE, `total` TIENE QUE SALIR DE `count="exact"` DE LA MISMA QUERY. Con
    # `.range(...)` puesto y esto sin tocar, `total` pasa a ser el largo de la página y miente sin
    # error: el chequeo de límite del export lo lee y va a creer que entra siempre.
    # Los tres pasos de la migración están en `services/_paginacion.py`.
    total: int
    page: int = 1
    page_size: int = 0
    total_pages: int = 0
